Events part 5: Create event page

darrenvong opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Includes a new page for creating a new event. More of #541. Almost there...

**Desktop**
![Screenshot 2021-08-07 at 03 50 53](https://user-images.githubusercontent.com/13669362/128585635-1ec841af-01e2-473d-bda3-9244775f7bd3.png)

**Mobile**
![Screenshot 2021-08-07 at 03 51 17](https://user-images.githubusercontent.com/13669362/128585653-a05723ce-3328-430a-bdff-879219280d47.png)
![Screenshot 2021-08-07 at 03 51 29](https://user-images.githubusercontent.com/13669362/128585654-7986663d-ed57-4676-aee4-a43422a79c5c.png)

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->
**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

darrenvong committed:
feat: add new events into service

Plus event feature behind feature flag

darrenvong committed:
feat: add time picker component

darrenvong committed:
feat: build out basic structure of create event form

darrenvong committed:
feat: update Timepicker to use utils/dayjs

darrenvong committed:
feat: add/link up create event mutation to page

Bit scrappy/hacky, more work needs to be done to make it more flexible

darrenvong committed:
feat: bake in transfer event in create event

Very hacky, since it's not flexible to allow different community ID.
Also no way to get community ID when the event is offline...

darrenvong committed:
feat: Use Switch instead of fragment for routes behind feature flag

React router seems to get confused and render everything that matches
the route otherwise...

darrenvong committed:
refactor: LocationAutocomplete to take in a label

darrenvong committed:
refactor: colocate constants, move create event page out of feature flag

darrenvong committed:
fix: location autocomplete test from refactor

darrenvong committed:
feat: add image upload to create event form

darrenvong committed:
refactor: decouple event form from the page component

So the form can be re-used by edit event later on

darrenvong committed:
fix: set event image to display as cover when one is available

darrenvong committed:
fix: event page cover photo styling

darrenvong committed:
feat: add unit tests for event form

darrenvong committed:
feat: add create event button link to community page

darrenvong committed:
feat: make location field required

Modified component to allow this to happen

darrenvong commented on app/frontend/src/components/Timepicker/Timepicker.tsx:
```diff
@@ -0,0 +1,61 @@
+  errorText?: string;
+}
+
+export default function Timepicker({
```
I'm a bit lost ~~as to why this doesn't work that well at all on keyboard. Tried changing the value back to using JavaScript `Date` instead and that's improved keyboard usability slightly,~~ never mind, that was actually unrelated. Following their example of adding a mask actually screwed it up more! But for some reason you still can't switch to "PM" without opening up the dialog :/ but in their demo it works...

darrenvong replied:
Also the clock seems to only allow drag and drop rather than clicking on the numbers, which makes this pretty difficult/probably less worthwhile to unit test. The author of RTL seems to suggest that too (and recommends doing this in Cypress e2e tests, which we currently don't have yet) https://github.com/testing-library/user-event/issues/440#issuecomment-685010755

lucaslcode replied:
This whole picker is a hot mess. I don't like the material time picker anyway - can we just use [`<input type="time">`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/time)? It's well supported, falls back to text input, and we can add format validation with the message "Must be a valid time in 24-hour format HH:MM" in case it does fall back.

darrenvong replied:
It would make things a lot simpler technically, but I feel like since we are using the Material date picker, having that side by side makes the UI looks quite inconsistent. Might it be worth using the combined time and date picker instead? Not sure if it suffers the same keyboard usability problem I'm having here with the time picker though.

darrenvong committed:
feat: add light "integration" test for create event page

Since most cases are covered in the event form tests

darrenvong committed:
fix: link styling from theme updates

It's pinned to the wrong font size since the recent token update

darrenvong committed:
feat: remove duplicate event page routing code

lucaslcode commented on app/frontend/src/features/communities/events/EventsList.tsx:
```diff
@@ -50,6 +56,13 @@ export default function EventsList({ community }: EventsListProps) {
   return (
     <>
       <SectionTitle icon={<CalendarIcon />}>{EVENTS_TITLE}</SectionTitle>
+      <Button
+        component={Link}
+        className={classes.createEventButton}
+        to={newEventRoute}
+      >
```
Can this be the same styling (should probably be shared) as the "Create discussion" button on the discussions listing page

lucaslcode commented on app/frontend/src/features/communities/events/CreateEventPage.tsx:
```diff
@@ -0,0 +1,95 @@
+          photoKey: data.eventImage,
+          startTime: finalStartDate,
+          endTime: finalEndDate,
+          // TODO: not hardcode this and allow user to specify community ID?
+          parentCommunityId: 1,
```
Until we do this or decide on something, I think add a little subtitle to the checkbox "Virtual events will be added to the global community"

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+
+  return (
+    <>
+      <ImageInput
+        alt={EVENT_IMAGE_INPUT_ALT}
+        className={classes.imageInput}
+        control={control}
+        id="event-image-input"
+        name="eventImage"
+        type="rect"
+      />
```
I think this needs a little pencil symbol somewhere to make it more obvious. But that can be another issue/pr.

darrenvong replied:
Where would you put the pencil without it looking a bit out of place? The way it is in the design would require some refactoring of the image input/possibly re-working it altogether which I don't fancy doing in this PR :p

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+        <Alert severity="error">
+          {error?.message ||
+            errors.eventImage?.message ||
+            // @ts-expect-error
```
add a reason - "react-hook-form typing incorrect for nested types"

darrenvong replied:
I'll just write for the comment for this first `ts-expect-error`. The rest follows the same pattern so I don't think the comment needs to be repeated all over the place

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+        <div className={classes.duoContainer}>
+          <Datepicker
+            control={control}
+            // @ts-expect-error
```
as above, and the others

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+          variant="standard"
+        />
+        <div className={classes.duoContainer}>
+          <Datepicker
```
I think the default start and end date should be tomorrow (arguable), and the default time should be rounded to the nearest hour.

Also, when you change the date, the end date should also change to the same interval apart (ie. if the dates are 2021-01-01 and 2021-01-03 and the start date is changed to 2021-01-03, the end date should change to 2021-01-05).
Same for the end time.

Maybe we need to add a `onPostChange` callback to call `setValue` in?

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+              variant="standard"
+            />
+          ) : (
+            <LocationAutocomplete
+              control={control}
+              fullWidth
+              label={LOCATION}
+              required={LOCATION_REQUIRED}
```
This is currently stripping parts of the address for the profile editing. Probably needs a prop to disable this.

darrenvong replied:
Hmm interesting, considering I barely changed the implementation of this apart from adding an optional required prop, so it should be the same as before if nothing gets passed in that prop like in all other places.

What locations were you typing in? I can't seem to reproduce it on the preview build

lucaslcode replied:
It is the same as before, but before it strips the street number and name from the address (right?). I tried something like 432 12th Street New York

darrenvong replied:
Oh I misread this and see what you mean now. The search results that come up here do only seem to show the stripped versions of the address, so yeah needs some tweaking...

lucaslcode commented on app/frontend/src/features/communities/events/CreateEventPage.tsx:
```diff
@@ -0,0 +1,95 @@
+      onMutate({ parentCommunityId }) {
+        return { parentCommunityId };
+      },
+      onSuccess(_, __, context) {
```
redirect to the event on success

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+        />
+        <div className={classes.duoContainer}>
+          <Datepicker
+            control={control}
```
This probably needs a rules prop and we should verify that the event is in the future and the end is after the start. The backend gives the error but better catch it early.

lucaslcode commented on app/frontend/src/features/communities/events/EventForm.tsx:
```diff
@@ -0,0 +1,257 @@
+          <Typography id="content-label" variant="h3" component="p">
+            {EVENT_DETAILS}
+          </Typography>
+          <MarkdownInput
+            control={control}
+            id="content"
+            name="content"
+            labelId="content-label"
+          />
```
We should add a required prop to catch the error before the backend.

lucaslcode commented on app/frontend/src/features/communities/events/CreateEventPage.tsx:
```diff
@@ -0,0 +1,95 @@
+          address: data.location.simplifiedName,
+          lat: data.location.location.lat,
+          lng: data.location.location.lng,
+        };
```
missing parent community id?

darrenvong replied:
I think for offline events, the backend has some magic which figures out which community ID to assign the event to based on location, so it's not required here

lucaslcode replied:
Looking at the backend code, it's optional and if you don't include it, it automatically assigns a community as you said. But I think we definitely want to specify here - I thought something was broken because I was using the global community to create a physical event in New York, and then I couldn't see it. But I just found it in the NY community.

darrenvong replied:
How do we feel about the possibility of letting people create an event that is clearly not in the community though? E.g. they could theoretically create an offline event in Paris in the New York community.

Otherwise, it's better for the cache invalidation too as it currently has to invalidate all community events keys due to not knowing the specific community it's in... so no complaints to do it technically this way!

darrenvong replied:
In addition to the above question... doing this is would be a bit clunky at the moment under `/event/new`, since we don't know the community ID the user comes from if they go to the URL directly...

Should we not have the route and do the `#new` thing like we did for discussions instead?

lucaslcode replied:
I think creating events outside the community is either fine or should be checked in the backend.

I would just do `/event/new/:communityId?` and if one isn't provided, do the automatic way.

darrenvong replied:
> if one isn't provided, do the automatic way.

As in, let the backend decide which community to put the event at as before?

darrenvong replied:
I've done it with route state, not put it under community wrapper though but I think that's ok if the event is being created for the community. It's closer to the design and `/event/new` could be visited standalone by going to the URL directly, so I think it makes sense to not have the wrapper to make things simpler.

lucaslcode reviewed (changes requested):
Creating a physical event didn't work for me on preview. I go the success message but couldn't find the event after - even after clearing the react-query cache. Is it because the parent community id is missing? Maybe that's also a bug in the backend then - accepting a parent community id of 0? @aapeliv 

I haven't looked thoroughly at the code yet, but it seems largely good, will continue.

lucaslcode commented on app/frontend/src/features/communities/constants.ts:
```diff
@@ -62,5 +63,9 @@ export const SEE_ALL_MODERATORS = "See all Community Builders";
 export const UNKNOWN_USER = "Unknown user";
+export const UPLOAD_IMAGE = "Upload image";
+export const UPLOAD_IMAGE_INSTRUCTIONS =
+  "Use the following markdown to display this image:";
```
I think this can actually be removed

lucaslcode commented on app/frontend/src/features/communities/events/EventPage.tsx:
```diff
@@ -218,7 +230,7 @@ export default function EventPage() {
                       {VIRTUAL_EVENT}
                     </Typography>
-                    <Link href={event.onlineInformation.link}>
+                    <Link href={event.onlineInformation.link} variant="body1">
```
Does this make a difference?

darrenvong replied:
I swear it did before, not anymore probably ~~because of the theme changes from the donation PR...~~ That's a lie, it's because I added this to the theme as a default prop to use if one isn't specified. I don't see why link text should be smaller than `body1` so I think that's ok.

Removed here!

lucaslcode commented on app/frontend/src/service/events.ts:
```diff
@@ -86,3 +90,55 @@ export async function setEventAttendance({
+  req.setContent(input.content);
+  req.setStartTime(Timestamp.fromDate(input.startTime));
+  req.setEndTime(Timestamp.fromDate(input.endTime));
+  req.setTimezone("UTC");
```
Why UTC? I think this should be empty/not set: `// the tzdata timezone identifier, empty string means tz of creating user`

darrenvong replied:
I'm a little confused about this tbh, since `Timestamp.fromDate` always converts thing back to UTC so that's why I set it... but maybe it's fine to leave it out? 🤷  The frontend also doesn't use it at the moment, it converts the `startTime`/`endTime` from `Timestamp` back to a `Dayjs` object and figures out the correct local time to display there (currently with the user's browser timezone, but could be updated to use the user profile's timezone).

So this separate timezone field feels a bit redundant?

darrenvong committed:
feat: share spacing styling for createX buttons

darrenvong committed:
refactor: cleanups from PR feedback

darrenvong committed:
feat: switch create event form to use native time picker

More accessible, easier to implement  technically & works fine with MUI
input so it's a no brainer

darrenvong commented:
Lots of good comments here, I've only partially addressed some of them so this is kinda back to a draft PR state again since there will be broken stuff again. Will look to address them in the coming days...

darrenvong committed:
refactor: start/end time into EventTimeChanger comp

To reduce re-rendering to those components only when things change

darrenvong committed:
feat: add more form validations before allowing event creation

darrenvong committed:
feat: allow location autocomplete to be configured to show full address

darrenvong committed:
feat: improve how errors are surfaced

For event details and location field

darrenvong committed:
test: event time changer logic

darrenvong committed:
feat: create event for community if ID known from navigation state

darrenvong commented:
Uhh I tried running the tests with `test-ci` locally with lots of workers and the tests that repeatedly fail seems to be ones we know about already, so hard to say whether anything I've written here is also too slow for GitLab...

darrenvong commented:
Otherwise, ready to be looked at again

darrenvong committed:
try run in band again

darrenvong committed:
debug: remove EventForm and EventTimeChanger tests

darrenvong committed:
debug: add EventForm test back

darrenvong committed:
debug: update EventTimeChanger tests to not go through MUI picker

darrenvong committed:
feat: add clarifying helper text for event image upload

darrenvong commented:
Assuming the tests will be green (finally, looks like the multiple datepicker clicks were too slow for CI lol), ready to be looked at again now @lucaslcode

lucaslcode reviewed (dismissed):
LGTM

Perhaps if using route state, it would be a good idea to test submission with route state and without? Up to you though

darrenvong committed:
fix: use full location name and optional chain route state access

darrenvong commented:
> Perhaps if using route state, it would be a good idea to test submission with route state and without? Up to you though

Good shout, I got lazy because I couldn't get the `LocationAutocomplete` thing working... turned out it was the fake timers messing with it. Glad I looked into it more as the tests caught some bug that'd have slipped in 😅

darrenvong commented on app/frontend/src/features/communities/events/CreateEventPage.test.tsx:
```diff
@@ -75,3 +72,96 @@ describe("Create event page", () => {
+      title: "Test event",
+      content: "sick social!",
+      photoKey: "",
+      startTime: expect.any(Date),
+      endTime: expect.any(Date),
```
msw and fakeTimers don't work well together - advancing the timer was causing all sort of weird React warnings 😢  so the check here sadly have to be more generic as I couldn't use fake timer here to lock the Date. At least the more specific check is done with the online event test above though, which is shared logic so I think this will do...
